applications/StructuralMechanicsApplication/custom_constitutive/linear_j2_plasticity_3d.cpp

```diff
@@ -209,48 +273,67 @@ void LinearJ2Plasticity3D::CalculateMaterialResponseCauchy(Parameters& rValues)
             CalculateTangentTensor(dgamma, norm_dev_stress, yield_function_normal_vector,
                                    rMaterialProperties, tangent_tensor);
         }
-
-    // Linear + exponential hardening
-    mStrainEnergy = 0.5 * inner_prod(strain_vector - mPlasticStrain, prod(elastic_tensor, strain_vector - mPlasticStrain))
-                    + GetPlasticPotential(rMaterialProperties);
     }
 }
 
-double& LinearJ2Plasticity3D::CalculateValue(Parameters& rParameterValues,
-        const Variable<double>& rThisVariable, double& rValue) {
-
-    //const Properties& MaterialProperties = rParameterValues.GetMaterialProperties();
-    //Vector& StrainVector = rParameterValues.GetStrainVector();
-    //Vector& StressVector = rParameterValues.GetStressVector();
-    //const double& E = MaterialProperties[YOUNG_MODULUS];
-    //const double& NU = MaterialProperties[POISSON_RATIO];
+//************************************************************************************
+//************************************************************************************
 
+double& LinearJ2Plasticity3D::CalculateValue(
+    Parameters& rParameterValues,
+    const Variable<double>& rThisVariable,
+    double& rValue
+    )
+{
     if(rThisVariable == STRAIN_ENERGY){
-        //TODO (marcelo): calculate STRAIN_ENERGY here and remove mStrainEnergy
-        //CalculateCauchyGreenStrain(rParameterValues, StrainVector);
-        //CalculatePK2Stress( StrainVector, StressVector, E, NU );
-        //rValue = 0.5 * inner_prod(StrainVector,StressVector); // Strain energy = 0.5*E:C:E
-        rValue = mStrainEnergy;
+        Vector& strain_vector = rParameterValues.GetStrainVector();
+        if (rParameterValues.GetProcessInfo().Has(INITIAL_STRAIN)) {
+            noalias(strain_vector) += rParameterValues.GetProcessInfo()[INITIAL_STRAIN];
+        }
+        const Properties& r_material_properties = rParameterValues.GetMaterialProperties();
+        Matrix elastic_tensor(6, 6);
```

Comment:
Since you know th size this could be a bounded_matrix I think

Reply:
I tried to change to bounded_matrix as per your suggestion, but I am getting type problems. I'd rather leave this for the next optimization stage, unless you @loumalouomega want to give it a try?